circuit_breaker.go

```diff
@@ -0,0 +1,159 @@
+// - Policies: Count5xxPolicy
+func NewCircuitBreaker() *CircuitBreaker {
+	return &CircuitBreaker{
+		policies:         []CircuitBreakerPolicy{Count5xxPolicy},
+		timeout:          10 * time.Second,
+		failThreshold:    3,
+		successThreshold: 1,
+	}
+}
+
+// SetPolicies sets the CircuitBreakerPolicy's that the CircuitBreaker will use to determine whether a response is a failure.
+func (cb *CircuitBreaker) SetPolicies(policies []CircuitBreakerPolicy) *CircuitBreaker {
+	cb.policies = policies
+	return cb
+}
+
+// SetTimeout sets the timeout duration for the CircuitBreaker
+func (cb *CircuitBreaker) SetTimeout(timeout time.Duration) *CircuitBreaker {
+	cb.timeout = timeout
+	return cb
+}
+
+// SetFailThreshold sets the number of failures that must occur within the timeout duration for the CircuitBreaker to
+// transition to the Open state.
+func (cb *CircuitBreaker) SetFailThreshold(threshold int) *CircuitBreaker {
+	cb.failThreshold = threshold
+	return cb
+}
+
+// SetSuccessThreshold sets the number of successes that must occur to transition the CircuitBreaker from the Half-Open state
+// to the Closed state.
+func (cb *CircuitBreaker) SetSuccessThreshold(threshold int) *CircuitBreaker {
+	cb.successThreshold = threshold
+	return cb
+}
+
+// CircuitBreakerPolicy is a function that determines whether a response should trip the circuit breaker
+type CircuitBreakerPolicy func(resp *http.Response) bool
+
+// Count5xxPolicy is a CircuitBreakerPolicy that trips the circuit breaker if the response status code is 500 or greater
+func Count5xxPolicy(resp *http.Response) bool {
+	return resp.StatusCode > 499
+}
+
+var ErrCircuitBreakerOpen = errors.New("rest: circuit breaker open")
+
+type circuitBreakerState uint32
+
+const (
+	closed circuitBreakerState = iota
+	open
+	halfOpen
+)
+
+func (cb *CircuitBreaker) getState() circuitBreakerState {
+	return circuitBreakerState(atomic.LoadUint32((*uint32)(&cb.state)))
+}
+
+func (cb *CircuitBreaker) allow() error {
+	if cb == nil {
+		return nil
+	}
+
+	if cb.getState() == open {
+		return ErrCircuitBreakerOpen
+	}
+
+	return nil
+}
+
+func (cb *CircuitBreaker) applyPolicies(resp *http.Response) {
+	if cb == nil {
+		return
+	}
+
+	failed := false
+	for _, policy := range cb.policies {
+		if policy(resp) {
+			failed = true
+			break
+		}
+	}
+
+	if failed {
+		if cb.failCount > 0 && time.Since(cb.lastFail) > cb.timeout {
+			cb.failCount = 0
+		}
+
+		switch cb.getState() {
+		case closed:
+			cb.failCount++
+			if cb.failCount == cb.failThreshold {
+				cb.open()
+			} else {
+				cb.lastFail = time.Now()
+			}
+		case halfOpen:
+			cb.open()
+		}
+	} else {
+		switch cb.getState() {
+		case closed:
+			return
+		case halfOpen:
+			cb.successCount++
+			if cb.successCount == cb.successThreshold {
+				cb.changeState(closed)
+			}
+		}
+	}
+
+	return
+}
+
+func (cb *CircuitBreaker) open() {
+	cb.changeState(open)
+	go func() {
+		time.Sleep(cb.timeout)
+		cb.changeState(halfOpen)
+	}()
+}
```

Comment:
Let's free the resource when the http connection is closed.

```suggestion
func (cb *CircuitBreaker) open(ctx context.Context) {
	cb.changeState(open)
	go func() {
		select {
			case <- time.After(cb.timeout):
				cb.changeState(halfOpen)
			case <-ctx.Done():
				// maybe the state could be reset here
		}
	}()
}
```

Reply:
Not sure I understand the reasoning here, correct me if I'm wrong:

There's no request related resource held here. Even if the http connection is closed, we should still change the state to halfOpen upon `timeout`. The CB belongs to the client, not to a specific request context.

Reply:
Oh, you are right.